tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,439 @@
+    
+    remote_op_relations = _get_remote_op_relations(graph_def,
+                                                   node_name_to_node_def,
+                                                   node_name_to_input_names)
+    
+    execution_specs = _get_execution_specs(graph_def,
+                                           graph,
+                                           node_name_to_node_def,
+                                           node_name_to_input_names,
+                                           remote_op_relations,
+                                           outputs)
+    # DEBUG
+    _debug_print_execution_specs(execution_specs)
+    
+    return execution_specs
+
+
+def _debug_print_execution_specs(execution_specs):
+    for execution_spec in execution_specs:
+        print('\nIs the current spec describing a remote op?', execution_spec['is_remote_op'])
+        print('Inputs:', execution_spec['inputs'])
+        print('Outputs:', execution_spec['outputs'])
+        print('Body nodes:', execution_spec['body_nodes'])
+
+
+def _get_graph(graph_def):
+    temp = tf.Graph()
+    with temp.as_default():
+        tf.import_graph_def(graph_def)
+        return tf.compat.v1.get_default_graph()
+
+
+def _get_node_name_to_node_def(graph_def):
+    return {node.name: node for node in graph_def.node}
+
+
+def _get_node_name_to_input_names(graph_def):
+    return {node.name: list(node.input) for node in graph_def.node}
+
+
+def _get_remote_op_relations(graph_def, node_name_to_node_def, node_name_to_input_names):
+    """Get {remote op: a list of remote op children},
+      
+    These remote op children must be executed before executing the remote op.
+    """
+    remote_op_relations = {}
+    
+    for node in graph_def.node:
+        if _check_remote_op(node):
+            remote_op_relations[node.name] = _bfs_get_remote_op_children(node.name,
+                                                                         node_name_to_node_def,
+                                                                         node_name_to_input_names)
+    return remote_op_relations
+
+
+def _bfs_get_remote_op_children(remote_op_name, node_name_to_node_def, node_name_to_input_names):
+    """Find the remote op children for a remote op"""
+    queue = [remote_op_name]
+    visited = set([remote_op_name])
+    remote_op_children = []
+    
+    while queue:
+        current_node_name = queue[0]
+        del queue[0]
+        
+        for input_node_name in node_name_to_input_names[current_node_name]:
+            if input_node_name not in visited:
+                visited.add(input_node_name)
+                input_node = node_name_to_node_def[input_node_name]
+                
+                if _check_remote_op(input_node):
+                    remote_op_children.append(input_node_name)
+                else:
+                    queue.append(input_node_name)
+    
+    return remote_op_children
+
+
+def _check_placeholder_op(node):
```

Comment:
nit: `_is_placeholder_op`

Reply:
Yes, this name is more appropriate. Changed